QuickJournalSync

papafe opened the pull request:
Example sync app that shows also:
- how to get a realm offline
- how to subscribe to changes in the session connection state
- how to create a callback for sync errors
- how to create a callback for subscriptions errors
- how to create a callback for client reset

It is very similar to the QuickJournal app that we already have plus:
- A login page (same as for the template app)
- On the main page there have been added:
	- A label that shows the connection state
	- A button to simulate session error
	- A button to simulate subscription error
	- A button to simulate client reset (that for now opens a dialog saying to terminate and re-enable sync)
- A `RealmService` has been added as the manager of everything realm related

Fixes #3327 and #3326

##  TODO

* [ ] Move to the realm-dotnet-sample repo
* [x] Adding more comments
* [x] Finish writing the README (including "merging" https://github.com/realm/realm-dotnet/pull/3394 somehow) 
* [x] Cleaning up

nirinchev commented on examples/QuickJournalSync/AppShell.xaml.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        InitializeComponent();
+
+        Routing.RegisterRoute("entryDetail", typeof(EntryDetailPage));
```
I know very little about AppShell, but why are the two other pages registered in xaml, and this one is in the code-behind?

papafe replied:
The pages that are registered here are [global routes](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/shell/navigation#register-detail-page-routes), they can be theoretically accessed from anywhere in the app. That usually makes sense for "detail" pages, that could be opened from multiple locations.

nirinchev commented on examples/QuickJournalSync/Messages/EntryModifiedMessage.cs:
```diff
@@ -0,0 +1,12 @@
+﻿using CommunityToolkit.Mvvm.Messaging.Messages;
+using QuickJournalSync.Models;
+
+namespace QuickJournalSync.Messages
```
Can move this to file-scoped namespace (also in the other files as well)

```suggestion
namespace QuickJournalSync.Messages;
```

papafe replied:
Done

nirinchev commented on examples/QuickJournalSync/Models/JournalEntry.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public string? Body { get; set; }
+
+        public DateTimeOffset CreatedDate { get; set; }
```
Should we make this private and computed automatically?

papafe replied:
We can compute this in the constructor

nirinchev replied:
I meant the setter 😛

papafe replied:
That we could 😁

nirinchev commented on examples/QuickJournalSync/README.md:
```diff
@@ -0,0 +1,386 @@
+# QuickJournalSync
+
+**QuickJournalSync** shows how to integrate Realm and [Atlas Device Sync](https://www.mongodb.com/docs/atlas/app-services/sync/get-started/) in a sample MAUI application, with particular focus on error handling, connection state changes and offline realms. This app is a synced version of [`QuickJournal`](https://github.com/realm/realm-dotnet/tree/main/examples/QuickJournal), so check that if you  ...
```
```suggestion
**QuickJournalSync** shows how to integrate Realm and [Atlas Device Sync](https://www.mongodb.com/docs/atlas/app-services/sync/get-started/) in a sample MAUI application, with particular focus on error handling, connection state changes and offline realms. This app is a synced version of [`QuickJournal`](https://github.com/realm/realm-dotnet/tree/main/examples/QuickJournal), so check that out if you are mainly interested in how Realm can be used effectively in conjunction with MVVM and data binding.
```

nirinchev commented on examples/QuickJournalSync/README.md:
```diff
@@ -0,0 +1,386 @@
+
+#### How the realm is opened in the sample application
+
+The way that the sample application deals with offline realms loosely follow the snippet for the recommended flow that was shown in the previous section:
```
```suggestion
The way that the sample application deals with offline realms loosely follows the snippet for the recommended flow that was shown in the previous section:
```

nirinchev commented on examples/QuickJournalSync/README.md:
```diff
@@ -0,0 +1,386 @@
+
+The way that the sample application deals with offline realms loosely follow the snippet for the recommended flow that was shown in the previous section:
+- When the application opens, the `LoginPage` is the first to be loaded.
+- If the user is not logged in already (`CurrentUser == null`), then the user can specify its email and password and login. The login method (`RealmService.LoginAsync`) uses `Realm.GetInstanceAsync`, because we want to be sure to have a completely synchronized realm before showing the main page to the user:
```
```suggestion
- If the user is not logged in already (`CurrentUser == null`), then the user can specify their email and password and login. The login method (`RealmService.LoginAsync`) uses `Realm.GetInstanceAsync`, because we want to be sure to have a completely synchronized realm before showing the main page to the user:
```

nirinchev commented on examples/QuickJournalSync/README.md:
```diff
@@ -0,0 +1,386 @@
+
+## Connection state changes
+
+In order to react to connection state changes for the synchronization you need to subscribe to `realm.SyncSession.PropertyChanges` and look out for the `Session.ConnectionState` property, like in the following example:
```
```suggestion
In order to react to connection state changes for the synchronization you need to subscribe to `realm.SyncSession.PropertyChanged` and look out for the `Session.ConnectionState` property, like in the following example:
```

nirinchev commented on examples/QuickJournalSync/README.md:
```diff
@@ -0,0 +1,386 @@
+
+### Simulate connection state changes in the app
+
+You can simulate connection state changes in the app by disabling the internet connection of the testing device, for instance. The app will show a toast when the connection state changes, as well as show the current connection state in the bottom part of the main page.
```
Can we put a button to do that? It's kind of difficult to shut down the internet on a simulator, so having a button that does session pause/resume could be nice.

papafe replied:
Just did it, and modified the README accordingly

nirinchev commented on examples/QuickJournalSync/README.md:
```diff
@@ -0,0 +1,386 @@
+There are different kind of errors that can happen when working on a synced application:
+- [Subscription errors](#subscription-errors), that are raised when there are issues with subscriptions in Flexible Sync;
+- [Session errors](#session-errors), that are raised when there is an issue with the synchronization or the connection;
+- [Client resets](#client-reset), that are raised in exceptional cases when a client cannot sync data with the backend.
```
```suggestion
- [Client reset errors](#client-reset), that are raised in exceptional cases when a client cannot sync data with the backend.
```

nirinchev commented on examples/QuickJournalSync/README.md:
```diff
@@ -0,0 +1,386 @@
+
+Subscription errors can happen when there is an issue with the subscriptions used in Flexible Sync, for example when trying to create a subscription for a property that is not in the list of [queryable fields](https://www.mongodb.com/docs/atlas/app-services/sync/configure/sync-settings/#queryable-fields). In the .NET SDK subscription errors are represented by the `SubscriptionException` class. 
+
+When there is an error with the subscriptions, then `realm.Subscriptions.State` will be equal to `SubscriptionSetState.Error` and the corresponding exception will be set on `realm.Subscriptions.Error`. In general, after updating the subscription set, it is convenient to wait on on `query.SubscribeAsync` to catch possible exceptions. Please note that if you are using the the old subscription API,  ...
```
```suggestion
When there is an error with the subscriptions, then `realm.Subscriptions.State` will be equal to `SubscriptionSetState.Error` and the corresponding exception will be available via the `realm.Subscriptions.Error` property. In general, after updating the subscription set, it is convenient to wait on `query.SubscribeAsync` to catch possible exceptions. Please note that if you are using the the old subscription API, then you will need to wait on `realm.Subscriptions.WaitForSynchronizationAsync`.
```

nirinchev commented on examples/QuickJournalSync/Services/DialogService.cs:
```diff
@@ -0,0 +1,44 @@
+{
+    public static class DialogService
+    {
+        private static Page? MainPage => Application.Current?.MainPage;
```
```suggestion
        private static Page MainPage => Application.Current?.MainPage ?? throw new InvalidOperationException("Cannot show dialogs without a Main Page!");
```

And then we don't need the `CheckIfMainPAgeAvailable()` method.

nirinchev commented on examples/QuickJournalSync/Services/RealmService.cs:
```diff
@@ -0,0 +1,261 @@
+
+        public static Realm GetMainThreadRealm()
+        {
+            lock (_mainRealmLock)
```
Why do you need this lock?

papafe replied:
We don't :)

nirinchev commented on examples/QuickJournalSync/Services/RealmService.cs:
```diff
@@ -0,0 +1,261 @@
+
+        public static async Task LoginAsync(string email, string password)
+        {
+        CheckIfInitialized();
+
+        await _app.LogInAsync(Credentials.EmailPassword(email, password));
+
+        // Creates a CancellationTokenSource that will be cancelled after 4 seconds.
+        var cts = new CancellationTokenSource(4000);
+
+        try
+        {
+            using var realm = await Realm.GetInstanceAsync(GetRealmConfig(), cts.Token);
+        }
+        catch (TaskCanceledException)
+        {
+            // If there are connectivity issues, or the synchronization is taking too long we arrive here
+        }
```
The indentation is weird here

nirinchev commented on examples/QuickJournalSync/Services/RealmService.cs:
```diff
@@ -0,0 +1,261 @@
+                }
+            }
+
+            return _mainThreadRealm;
```
Should we have some guard that asserts that we're indeed on the main thread? Doesn't seem like anything is guaranteeing this here.

papafe replied:
We can do that

nirinchev commented on examples/QuickJournalSync/Services/RealmService.cs:
```diff
@@ -0,0 +1,261 @@
+
+        private static void HandleSyncSessionPropertyChanged(object? sender, PropertyChangedEventArgs e)
+        {
+    if (sender is Session session && e.PropertyName == nameof(Session.ConnectionState))
+    {
+        // React to connection state changes
+        LogAndShowToast($"New connection state: {session.ConnectionState}");
+        SyncConnectionStateChanged?.Invoke(null, session.ConnectionState);
+    }
```
Indentation

nirinchev reviewed (commented):
Looks really nice. I mostly have cosmetic questions/suggestions.

papafe commented:
I've added the changes according to the comments
- file scoped namespaces
- added button to test connectivity changes
- spelling fixes
- removed lock
- add check to verify GetMainThreadRealm is called on main thread
- small fixes

Once this is approved, I'll just push it to master on the realm-dotnet-example repo

papafe commented:
Closing this has this been just merged in https://github.com/realm/realm-dotnet-samples
